[WebGPU] writeBuffer test is failing

mwyrzykowski opened the pull request:
#### c6a5bcca33e3147a0aaa5ea1f3aa2384aae383da
<pre>
[WebGPU] writeBuffer test is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=253748">https://bugs.webkit.org/show_bug.cgi?id=253748</a>
&lt;radar://106583393&gt;

Reviewed by Tadeu Zagallo.

Multiply offsets by the size of the element and fix race condition between
writeBuffer() and isIdle().

Also enable writeBuffer.html as it is passing but disable createSampler.html which
is a flaky 1-3% failure.

* LayoutTests/http/tests/webgpu/webgpu/api/operation/queue/writeBuffer-expected.txt:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::computeElementSize):
(WebCore::GPUQueue::writeBuffer):
* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebGPU/WebGPU/Queue.h:
(WebGPU::Queue::isIdle const): Deleted.
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::validateWriteBuffer const):
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::isIdle const):
(WebGPU::Queue::writeTexture):

Canonical link: <a href="https://commits.webkit.org/261802@main">https://commits.webkit.org/261802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9869f31630383c448e244ea821579e0b3fc2631

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112413 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21558 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1062 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4192 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121013 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116477 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22898 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12687 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5189 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118180 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17010 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100201 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105467 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98965 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/734 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46021 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13915 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/786 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/94967 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14614 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10135 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19958 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52787 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8235 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16422 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -67,14 +67,34 @@ void GPUQueue::onSubmittedWorkDone(OnSubmittedWorkDonePromise&& promise)
+    auto dataSize = data.length();
+    auto contentSize = optionalSize.has_value() ? (elementSize * optionalSize.value()) : (dataSize - dataOffset);
+
+    if (dataOffset > dataSize || dataOffset + contentSize > dataSize || (contentSize % 4))
```
Can `dataOffset + contentSize` overflow? Otherwise I don't think you need the first check, right?

mwyrzykowski replied:
It could overflow and I should check for that, thanks for catching. The overflow may be tested for in WebGPU.framework, I will double-check

mwyrzykowski replied:
Oh I know why the check is needed @tadeuzagallo 

contentSize = dataSize - dataOffset, if dataOffset > dataSize, then contentSize will overflow

mwyrzykowski replied:
There is also the issue that `(elementSize * optionalSize.value())` could overflow or `dataOffset + contentSize` could overflow. I will check the overflows

litherum commented on Source/WebCore/Modules/WebGPU/GPUQueue.h:
```diff
@@ -60,7 +60,7 @@ class GPUQueue : public RefCounted<GPUQueue> {
     void onSubmittedWorkDone(OnSubmittedWorkDonePromise&&);
 
-    void writeBuffer(
+    ExceptionOr<void> writeBuffer(
```
Is this correct? It throws, instead of using GPUErrorScope?

mwyrzykowski replied:
Yes that is correct:
```
If any of the following conditions are unsatisfied, throw [OperationError](https://webidl.spec.whatwg.org/#operationerror) and stop.

contentsSize ≥ 0.
dataOffset + contentsSize ≤ dataSize.
contentsSize, converted to bytes, is a multiple of 4 bytes.
```

The specification is quite confusing in this regard. Some things throw exceptions, other items use GPUErrorScope. I'm trying to understand why there are these two methods of error handling but perhaps a specification issue is warranted.

litherum commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
```diff
@@ -80,7 +80,7 @@ void RemoteQueueProxy::writeBuffer(
         return;
 
-    auto sendResult = send(Messages::RemoteQueue::WriteBuffer(convertedBuffer, bufferOffset, Vector<uint8_t>(static_cast<const uint8_t*>(source) + dataOffset, size.value_or(byteLength - dataOffset))));
+    auto sendResult = sendSync(Messages::RemoteQueue::WriteBuffer(convertedBuffer, bufferOffset, Vector<uint8_t>(static_cast<const uint8_t*>(source) + dataOffset, size.value_or(byteLength - dataOffset))));
```
Please don't do this :)

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.cpp:
```diff
@@ -75,21 +75,26 @@ void RemoteQueue::onSubmittedWorkDone(CompletionHandler<void()>&& callback)
     PAL::WebGPU::Size64 bufferOffset,
-    Vector<uint8_t>&& data)
+    Vector<uint8_t>&& data,
+    CompletionHandler<void()>&& callback)
```
I don't think this is right. Writing a buffer should just queue up the write operation. It doesn't need a callback, and it doesn't need to block the web process.

If the buffer isn't in use, and the buffer is shared, writeBuffer can just be a memcpy. If the buffer is in use, or if it isn't shared, writeBuffer should copy the data into a temporary buffer, and enqueue a blit command to copy the data from the temporary buffer to its final destination.

IIRC I already implemented this logic :/

mwyrzykowski replied:
Oh good point I misunderstood the CTS test.

Indeed I think this blocking should be removed. Instead I think we have a bug in Buffer::mapAsync.

mwyrzykowski replied:
Specifically `Buffer::getMappedRange` in WebGPU.framework returns `return static_cast<char*>(m_buffer.contents) + offset;` but if there is a queued up blit command on that buffer then that is wrong I think. It will return the buffer's current contents as opposed to the contents after the blit command finishes

litherum replied:
getMappedRange shouldn't block; instead, mapAsync's promise shouldn't resolve until all commands referencing the buffer are finished

mwyrzykowski replied:
mapAsync is also using `onSubmittedWorkDone()` 

My current theory is the bug is in `unmap`, but I am investigating.

litherum commented on Source/WebGPU/WebGPU/ExternalTexture.h:
```diff
@@ -0,0 +1,72 @@
+class Device;
+
+// https://gpuweb.github.io/gpuweb/#gpuexternaltexture
+class ExternalTexture : public WGPUExternalTextureImpl, public RefCounted<ExternalTexture> {
```
Did you mean to add this in this patch? This patch is about a failing buffer test, but this is a texture class...

mwyrzykowski replied:
Certainly not, I should have not git added that entire path :(

litherum commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -34,12 +34,35 @@
     , m_device(device)
 {
+    m_modifiedResources = [NSMutableSet set];
+    m_modifiedTextures = [NSMutableSet set];
```
Can we use `Vector<RetainPtr<MTLResource>>` instead? We generally prefer WTF types in WebKit.

mwyrzykowski replied:
I can use the WTF type but then to use HashSet I think `id<MTLTexture>` needs a hash function, it has one with `-[NSObject hash]` but I need to see how to have HashSet use that

mwyrzykowski replied:
I didn't want to use Vector to avoid searching for duplicates or encoding multiple synchronization calls

but I'm checking to see if these calls are even needed after fixing the other race condition

litherum commented on Source/WebGPU/WebGPU/Queue.h:
```diff
@@ -102,6 +106,8 @@ class Queue : public WGPUQueueImpl, public ThreadSafeRefCounted<Queue> {
     using OnSubmittedWorkDoneCallbacks = Vector<WTF::Function<void(WGPUQueueWorkDoneStatus)>>;
     HashMap<uint64_t, OnSubmittedWorkDoneCallbacks, DefaultHash<uint64_t>, WTF::UnsignedWithZeroKeyHashTraits<uint64_t>> m_onSubmittedWorkDoneCallbacks;
+    NSMutableSet<id<MTLResource>> *m_modifiedResources;
+    NSMutableSet<TextureSliceLevel*> *m_modifiedTextures;
```
I kind of think these should be in their own class, along with `m_blitCommandEncoder`. Is there a better refactoring we can do here, instead of having a bunch of sibling member variables with the same lifetime? If we make a new class just for these fields, we can give that class its own API (perhaps even using RAII), making it harder to accidentally not manage some state correctly.

I'm imagining something like
```
class TemporaryBlitEncoder {
public:
    void copySomethingToSomethingElse(Something, SomethingElse);
    ...
private:
    id<MTLBlitCommandEncoder> m_blitCommandEncoder { nil };
    using IThinkThereIsASpecTermForThisButIForgetWhatItIs = std::variant<RetainPtr<id<MTLResource>>, TextureSliceLevel>;
    HashSet<IThinkThereIsASpecTermForThisButIForgetWhatItIs> m_resources;
};
...
class Queue {
...
    std::optional<TemporaryBlitEncoder> m_temporaryBlitEncoder;
};
```

or something like that

mwyrzykowski replied:
Sounds good! It appears these changes are not needed, so I removed them for now

litherum commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -34,12 +34,35 @@
 #import "Texture.h"
 #import <wtf/CheckedArithmetic.h>
 
+@interface TextureSliceLevel : NSObject
```
Can we do this with C++ instead of Obj-C?

litherum commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -249,8 +291,12 @@ static bool validateWriteBufferInitial(size_t size)
         return;
+    }
+
+    didModifyResource(buffer.buffer());
```
I don't think this is right. We don't want to synchronize literally every buffer that gets written to. It will be way more common for buffers to be written to from the CPU, but never read from the CPU, and this patch would cause lots of needless synchronization in that case. I think mapAsync() should schedule a synchronize call if one is necessary, rather than doing it unconditionally here.

mwyrzykowski replied:
I think we still need the synchronization independent of mapAsync as the CTS test performs overlapping writes which can use memcpy or the blit encoder depending if the queue is idle or not.

mwyrzykowski replied:
I am looking at the documentation more closely this should only be necessary for MTLStorageModeManaged so I need to determine where the race condition is on Apple Silicon. The test causing the race is this one:
```
    for (const { bufferOffset, data, arrayType, useArrayBuffer, dataOffset, dataSize } of writes) {
      const TypedArrayConstructor = globalThis[arrayType];
      const writeData = new TypedArrayConstructor(data);
      const writeSrc = useArrayBuffer ? writeData.buffer : writeData;
      this.queue.writeBuffer(buffer, bufferOffset, writeSrc, dataOffset, dataSize);
      memcpy(
      { src: writeSrc, start: dataOffset, length: dataSize },
      { dst: expectedData, start: bufferOffset });

    }

    this.debug(`expectedData: [${expectedData.join(', ')}]`);
    this.expectGPUBufferValuesEqual(buffer, expectedData);
```

mwyrzykowski commented on LayoutTests/platform/mac/TestExpectations:
```diff
@@ -2456,7 +2456,7 @@ http/tests/webgpu/webgpu/api/validation/image_copy/buffer_related.html [ Skip ]
 http/tests/webgpu/webgpu/api/validation/image_copy/buffer_texture_copies.html [ Skip ]
 http/tests/webgpu/webgpu/api/validation/image_copy/texture_related.html [ Skip ]
 http/tests/webgpu/webgpu/api/validation/image_copy/layout_related.html [ Skip ]
-http/tests/webgpu/webgpu/api/validation/createSampler.html [ Pass ]
```
this was observed to be a flaky failure (1-3 failures out of 100 iterations), disabling until it is fixed

mwyrzykowski commented on Source/WebGPU/WebGPU/Queue.mm:
```diff
@@ -259,6 +263,11 @@ static bool validateWriteBufferInitial(size_t size)
 
+bool Queue::isIdle() const
+{
+    return m_submittedCommandBufferCount == m_completedCommandBufferCount && !m_blitCommandEncoder;
```
@litherum are you ok with this change? It fixed the race locally and EWS is green.

mwyrzykowski committed:
[WebGPU] writeBuffer test is failing
https://bugs.webkit.org/show_bug.cgi?id=253748
<radar://106583393>

Reviewed by Tadeu Zagallo.

Multiply offsets by the size of the element and fix race condition between
writeBuffer() and isIdle().

Also enable writeBuffer.html as it is passing but disable createSampler.html which
is a flaky 1-3% failure.

* LayoutTests/http/tests/webgpu/webgpu/api/operation/queue/writeBuffer-expected.txt:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::computeElementSize):
(WebCore::GPUQueue::writeBuffer):
* Source/WebCore/Modules/WebGPU/GPUQueue.h:
* Source/WebGPU/WebGPU/Queue.h:
(WebGPU::Queue::isIdle const): Deleted.
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::validateWriteBuffer const):
(WebGPU::Queue::writeBuffer):
(WebGPU::Queue::isIdle const):
(WebGPU::Queue::writeTexture):

Canonical link: https://commits.webkit.org/261802@main
